src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -587,11 +590,11 @@ final class LazyList[+A] private(private[this] var lazyState: () => LazyList.Sta
 
   @inline private def lazySlidingImpl(size: Int, step: Int): Iterator[LazyList[A]] =
     if (knownIsEmpty) Iterator.empty
-    else Iterator.empty ++ slidingImpl(size, step) // concat with empty iterator so that `slidingImpl` is lazy
+    else Iterator.empty ++ slidingImpl(size, step, size - step max 0) // concat with empty iterator so that `slidingImpl` is lazy
 
-  private def slidingImpl(size: Int, step: Int): Iterator[LazyList[A]] =
-    if (isEmpty) Iterator.empty
-    else Iterator.single(take(size)) ++ drop(step).slidingImpl(size, step)
+  private def slidingImpl(size: Int, step: Int, minLen: Int): Iterator[LazyList[A]] =
+    if (lengthLteq(minLen)) Iterator.empty
```

Comment:
basically, incomplete windows must be larger than `size - step`, for reasons that are probably obvious if you do the maths, but also fairly apparent if you run lines like the following in a REPL:

```scala
scala> List.from(1 to 7).sliding(3, 2).foreach(x => println(x))
List(1, 2, 3)
List(3, 4, 5)
List(5, 6, 7)

scala> List.from(1 to 6).sliding(3, 2).foreach(x => println(x))
List(1, 2, 3)
List(3, 4, 5)
List(5, 6)

scala> List.from(1 to 5).sliding(3, 2).foreach(x => println(x))
List(1, 2, 3)
List(3, 4, 5)
```

```scala
scala> List.from(1 to 7).sliding(4, 3).foreach(x => println(x))
List(1, 2, 3, 4)
List(4, 5, 6, 7)

scala> List.from(1 to 6).sliding(4, 3).foreach(x => println(x))
List(1, 2, 3, 4)
List(4, 5, 6)

scala> List.from(1 to 5).sliding(4, 3).foreach(x => println(x))
List(1, 2, 3, 4)
List(4, 5)

scala> List.from(1 to 4).sliding(4, 3).foreach(x => println(x))
List(1, 2, 3, 4)
```

```scala
scala> List.from(1 to 6).sliding(4, 2).foreach(x => println(x))
List(1, 2, 3, 4)
List(3, 4, 5, 6)

scala> List.from(1 to 5).sliding(4, 2).foreach(x => println(x))
List(1, 2, 3, 4)
List(3, 4, 5)

scala> List.from(1 to 4).sliding(4, 2).foreach(x => println(x))
List(1, 2, 3, 4)
```

Reply:
The termination condition is "stop when you hit nothing or after the first under-full window"?  Weird.  I never noticed that.  I have trouble envisioning for which problem this is the right behavior.  Probably too late to change it globally, though.